pkg/controllers/networkpolicy/cloudresource.go

```diff
@@ -31,148 +30,108 @@ const (
 )
 
 var (
-	resourceNPStatusSetter = map[securitygroup.CloudResourceType]func(tracker *cloudResourceNPTracker,
+	resourceNPStatusSetter = map[securitygroup.CloudResourceType]func(tracker *CloudResourceNpTracker,
 		reconciler *NetworkPolicyReconciler) (bool, error){
-		securitygroup.CloudResourceTypeVM: vmNPStatusSetter,
+		securitygroup.CloudResourceTypeVM: vmNpStatusSetter,
 	}
 )
 
 const (
 	AppliedSecurityGroupDeleteError = "Deleting/Detaching appliedTo sg %v: %v"
 )
 
-func vmNPStatusSetter(tracker *cloudResourceNPTracker, r *NetworkPolicyReconciler) (bool, error) {
-	log := r.Log.WithName("NPTracker")
-	status := tracker.computeNPStatus(r)
-	updated := false
-
-	vmItems, err := r.Inventory.GetVmFromIndexer(indexer.VirtualMachineByCloudId, tracker.cloudResource.Name)
-	if err != nil {
-		log.Error(err, "failed to get VM from VM cache")
-		return false, err
-	}
-	for _, item := range vmItems {
-		vm := item.(*runtimev1alpha1.VirtualMachine)
-		npStatus, ok := status[vm.Namespace]
-		if len(status[""]) > 0 {
-			if npStatus == nil {
-				npStatus = make(map[string]string)
-			}
-			for k, v := range status[""] {
-				npStatus[k] = v
-			}
+func vmNpStatusSetter(tracker *CloudResourceNpTracker, r *NetworkPolicyReconciler) (bool, error) { //nolint:unparam
+	status := tracker.computeNpStatus(r)
+	npStatus := status[tracker.NamespacedName.Namespace]
+	if len(status[""]) > 0 {
+		if npStatus == nil {
+			npStatus = make(map[string]string)
 		}
-		indexKey := types.NamespacedName{Namespace: vm.Namespace, Name: vm.Name}
-		obj, found, _ := r.virtualMachinePolicyIndexer.GetByKey(indexKey.String())
-		// no policy to update.
-		if !ok && !found {
-			continue
-		}
-
-		var cache *NetworkPolicyStatus
-		if found {
-			cache = obj.(*NetworkPolicyStatus)
-		} else {
-			cache = newNetworkPolicyStatus(indexKey.Namespace, indexKey.Name)
-		}
-		// policy status did not change.
-		if ok && reflect.DeepEqual(cache.NPStatus, npStatus) {
-			continue
-		}
-
-		// cache operation.
-		if len(npStatus) != 0 {
-			cache.NPStatus = npStatus
-			if err := r.virtualMachinePolicyIndexer.Update(cache); err != nil {
-				// mark dirty and retry later on error.
-				tracker.markDirty()
-				continue
-			}
-			log.V(1).Info("Update vmp status", "resource", cache.String(), "status", npStatus)
-		} else {
-			if err := r.virtualMachinePolicyIndexer.Delete(cache); err != nil {
-				tracker.markDirty()
-				continue
-			}
-			log.V(1).Info("Delete vmp status", "resource", cache.String())
+		for k, v := range status[""] {
+			npStatus[k] = v
 		}
-		updated = true
 	}
-	return updated, nil
+	// TODO: Change computeNpStatus to update the tracker NP status.
+	tracker.NPStatus = npStatus
```

Comment:
Here we update the tracker field and did not update the indexer. I assume it works because tracker returned by indexer is a pointer? Maybe we should add a comment noting this, or call the indexer update explicitly in `processCloudResourceNpTrackers`?

Reply:
yes you right. Modified the function processCloudResourceNpTrackers to update the indexers.